plugin/src/main/kotlin/tanvd/grazi/grammar/GrammarChecker.kt

```diff
@@ -7,30 +7,35 @@ import tanvd.kex.*
 
 class GrammarChecker(private val ignoreChar: LinkedSet<(CharSequence, Char) -> Boolean> = LinkedSet(),
                      private val replaceChar: LinkedSet<(CharSequence, Char) -> Char?> = LinkedSet(),
-                     private val ignoreToken: LinkedSet<(String) -> Boolean> = LinkedSet()) {
+                     private val ignoreToken: LinkedSet<(String) -> Boolean> = LinkedSet(),
+                     private val afterResultReplace: LinkedSet<(String) -> String> = LinkedSet()) {
 
     constructor(checker: GrammarChecker, ignoreChar: LinkedSet<(CharSequence, Char) -> Boolean> = LinkedSet(),
                 replaceChar: LinkedSet<(CharSequence, Char) -> Char?> = LinkedSet(),
-                ignoreToken: LinkedSet<(String) -> Boolean> = LinkedSet())
-            : this(LinkedSet(checker.ignoreChar + ignoreChar), LinkedSet(checker.replaceChar + replaceChar), LinkedSet(checker.ignoreToken + ignoreToken))
+                ignoreToken: LinkedSet<(String) -> Boolean> = LinkedSet(),
+                afterResultReplace: LinkedSet<(String) -> String> = LinkedSet())
+            : this(LinkedSet(checker.ignoreChar + ignoreChar), LinkedSet(checker.replaceChar + replaceChar),
+                LinkedSet(checker.ignoreToken + ignoreToken), LinkedSet(checker.afterResultReplace + afterResultReplace))
 
     companion object {
         object Rules {
-            val deduplicateBlanks: (CharSequence, Char) -> Boolean = { str, cur ->
-                str.lastOrNull()?.isWhitespace().orTrue() && cur.isWhitespace()
+            val deduplicateBlanks: (CharSequence, Char) -> Boolean = { prefix, cur ->
+                prefix.lastOrNull()?.isWhitespace().orTrue() && cur.isWhitespace()
             }
             //TODO probably we need to flat them only if previous char is not end of sentence punctuation mark
             val flatNewlines: (CharSequence, Char) -> Char? = { _, cur ->
                 Text.isNewline(cur).ifTrue { ' ' }
             }
 
-            val ignoreQuotesAtBorders: (CharSequence, Char) -> Boolean = { prev, cur ->
-                (cur == '\'' || cur == '\"') && (prev.isEmpty() || prev.last() == cur)
+            val ignoreQuotesAtBorders: (String) -> String = { result ->
+                val begin = result.indexOfFirst { !it.isWhitespace() && it != '\"' && it != '\'' }
+                val end = result.indexOfLast { !it.isWhitespace() && it != '\"' && it != '\'' }
+                "\r".repeat(begin) + result.substring(begin, end + 1) + "\r".repeat(result.length - end - 1)
```

Comment:
Will it work on windows? It looks like a really poor hack :(

Reply:
Let's try to understand what is the actual problem with quotes. If we need to unquote string we may add function that will remove parts of strings on borders and will align typos indices accordingly.

Reply:
Outdated